templates/components/agents/python/financial_report/app/workflows/financial_report.py

```diff
@@ -0,0 +1,300 @@
+        response = await chat_with_tools(
+            self.llm,
+            [
+                self.query_engine_tool,
+                self.code_interpreter_tool,
+                self.document_generator_tool,
+            ],
+            chat_history,
+        )
+        is_tool_call = isinstance(response, ToolCallResponse)
+        if not is_tool_call:
+            # If no tool call, return the response generator
+            return StopEvent(result=response.generator)
+        # calling different tools at the same time is not supported at the moment
+        # add an error message to tell the AI to process step by step
+        tool_calls = response.tool_calls
+        if is_calling_different_tools(tool_calls):
+            self.memory.put(
+                ChatMessage(
+                    role=MessageRole.ASSISTANT,
+                    content="Cannot call different tools at the same time. Try calling one tool at a time.",
+                )
+            )
+            return InputEvent(input=self.memory.get())
+        self.memory.put(response.tool_call_message)
+        tool_name = tool_calls[0].tool_name
+        match tool_name:
+            case self.query_engine_tool.metadata.name:
+                return ResearchEvent(input=tool_calls)
+            case self.code_interpreter_tool.metadata.name:
+                return AnalyzeEvent(input=tool_calls)
+            case self.document_generator_tool.metadata.name:
+                return ReportEvent(input=tool_calls)
+
+    @step()
+    async def research(self, ctx: Context, ev: ResearchEvent) -> AnalyzeEvent:
+        """
+        Do a research to gather information for the user's request.
+        A researcher should have these tools: query engine, search engine, etc.
+        """
+        ctx.write_event_to_stream(
+            AgentRunEvent(
+                name="Researcher",
+                msg="Starting research",
+            )
+        )
+        tool_calls = ev.input
+
+        tool_messages = await call_tools(
+            ctx=ctx,
+            agent_name="Researcher",
+            tools=[self.query_engine_tool],
+            tool_calls=tool_calls,
+        )
+        self.memory.put_messages(tool_messages)
+        # Request to analyze the research result
+        message = ChatMessage(
+            role=MessageRole.ASSISTANT,
+            content="I've finished the research. Please analyze the result.",
+        )
+        return AnalyzeEvent(input=message)
+
+    @step()
+    async def analyze(self, ctx: Context, ev: AnalyzeEvent) -> InputEvent:
+        """
+        Analyze the research result.
+        """
+        ctx.write_event_to_stream(
+            AgentRunEvent(
+                name="Analyst",
+                msg="Starting analysis",
+            )
+        )
+        event_requested_by_workflow_llm = isinstance(ev.input, list)
+        # Requested by the workflow LLM Input step, it's a tool call
+        if event_requested_by_workflow_llm:
```

Comment:
```suggestion
        if ev.has_tool_calls:
```

Reply:
that's the only thing a bit hacky (because it has an assumption about who is calling) - can't we send two different AnalyzeEvent events instead?